wgsl/index.bs

```diff
@@ -395,34 +418,127 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+
+TODO: Add requirement for *offset* attribute on struct members.
+
+### Storage Classes ### {#storage-class}
+
+Memory locations are partitioned into <dfn noexport>storage classes</dfn>.
+Each storage class has unique properties determining
+mutability, visibility, the values it may contain,
+and how to use variables with it.
+
+<table class='data'>
+  <thead>
+    <tr><td>Storage class
+        <td>Readable by shader?<br>Writable by shader?
+        <td>Sharing among invocations
+        <td>Variable scope
+        <td>Restrictions on stored values
+        <td>Notes
+  </thead>
+  <tr><td>input
+      <td>Read-only
+      <td>Same invocation only
+      <td>[=Module scope=]
+      <td>[=IO-shareable=]
+      <td>Input from an upstream pipeline stage, or from the implementation.
+  <tr><td>output
+      <td>Read-write
+      <td>Same invocation only
+      <td>[=Module scope=]
+      <td>[=IO-shareable=]
+      <td>Output to a downstream pipeline stage.
+  <tr><td>function
+      <td>Read-write
+      <td>Same invocation only
+      <td>[=Function scope=]
+      <td>[=Storable=]
+      <td>
+  <tr><td>private
+      <td>Read-write
+      <td>Same invocation only
+      <td>[=Module scope=]
+      <td>[=Storable=]
+      <td>
+  <tr><td>workgroup
+      <td>Read-write
+      <td>Invocations in the same [=workgroup=]
+      <td>[=Module scope=]
+      <td>[=Storable=]
+      <td>
+  <tr><td>uniform
+      <td>Read-only
+      <td>Invocations in the same [=shader stage=]
+      <td>[=Module scope=]
+      <td>[=Host-shareable=]
+      <td>The [=store type=] of a variable in the `uniform` storage class
+          must be a `block`-decorated [=host-shareable=] structure type.
+          See also [[#memory-layout]].
+  <tr><td>storage_buffer
+      <td>Readable.<br>
+          Also writable if the variable is not read-only.
+      <td>Invocations in the same [=shader stage=]
+      <td>[=Module scope=]
+      <td>[=Host-shareable=]
+      <td>The [=store type=] of a variable in the `storage_buffer` storage class
+          must be a `block`-decorated [=host-shareable=] structure type.
+          See also [[#memory-layout]].
+  <tr><td>texture
+      <td>Read-only for [=read-only texture=]<br>
+          Write-only for [=write-only texture=]
+      <td>Invocations in the same shader stage
+      <td>[=Module scope=]
+      <td>Depends on texture format
+      <td>
+  <tr><td>uniform_constant
+      <td>Read-only
+      <td>Invocations in the same shader stage
+      <td>[=Module scope=]
+      <td>Opaque representation of handle to a sampler or texture
+      <td>
+</table>
+
+Issue: The note about read-only `storage_buffer` variables may change depending
+on the outcome of https://github.com/gpuweb/gpuweb/issues/935
 
 <pre class='def'>
 storage_class
   : INPUT
   | OUTPUT
+  | FUNCTION
+  | PRIVATE
   | UNIFORM
   | WORKGROUP
   | UNIFORM_CONSTANT
   | STORAGE_BUFFER
   | IMAGE
```

Comment:
You renamed this to `texture` in the table below, should this also change to match?

Reply:
It turns out we would only need this to express a pointer-to-image.
But that's only needed if we do image atomics.

However, we've learned that WGSL won't support image atomics. (see https://github.com/gpuweb/gpuweb/issues/728#issuecomment-696896462 ).

So the best thing is to remove "image" storage class altogether.